File: java/yass-base/main/ch/softappeal/yass/core/Interceptor.java
package ch.softappeal.yass.core;

import ch.softappeal.yass.util.Nullable;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Proxy;
import java.util.Objects;

/**
 * Intercepts a method invocation.
 */
@FunctionalInterface public interface Interceptor {

    /**
     * Performs extra treatments before and after the invocation.
     * @param arguments see {@link InvocationHandler#invoke(Object, Method, Object[])}
     * @return {@link Invocation#proceed()}
     * @throws Exception exception of {@link Invocation#proceed()}
     */
    @Nullable Object invoke(Method method, @Nullable Object[] arguments, Invocation invocation) throws Exception;

    /**
     * Calls {@link Invocation#proceed()}.
     */
    Interceptor DIRECT = (method, arguments, invocation) -> invocation.proceed();

    static Interceptor composite(final Interceptor interceptor1, final Interceptor interceptor2) {
        Objects.requireNonNull(interceptor1);
        Objects.requireNonNull(interceptor2);
        if (interceptor1 == DIRECT) {
            return interceptor2;
        }
        if (interceptor2 == DIRECT) {
            return interceptor1;
        }
        return (method, arguments, invocation) -> interceptor1.invoke(
            method,
            arguments,
            () -> interceptor2.invoke(method, arguments, invocation)
        );
    }

    static Interceptor composite(final Interceptor... interceptors) {
        Interceptor composite = DIRECT;
        for (final Interceptor interceptor : interceptors) {
            composite = composite(composite, interceptor);
        }
        return composite;
    }

    static @Nullable Object invoke(final Interceptor interceptor, final Method method, final @Nullable Object[] arguments, final Object implementation) throws Exception {
        return interceptor.invoke(method, arguments, () -> {
            try {
                return method.invoke(implementation, arguments);
            } catch (final InvocationTargetException e) {
                try {
                    throw e.getCause();
                } catch (final Exception | Error e2) {
                    throw e2;
                } catch (final Throwable t) {
                    throw new Error(t);
                }
            }
        });
    }

    /**
     * @param <C> the contract type
     * @return a proxy for implementation using interceptors
     */
    static <C> C proxy(final Class<C> contract, final C implementation, final Interceptor... interceptors) {
        Objects.requireNonNull(implementation);
        final Interceptor interceptor = composite(interceptors);
        if (interceptor == DIRECT) {
            Objects.requireNonNull(contract);
            return implementation;
        }
        return contract.cast(Proxy.newProxyInstance(
            contract.getClassLoader(),
            new Class<?>[] {contract},
            (proxy, method, arguments) -> invoke(interceptor, method, arguments, implementation)
        ));
    }

    /**
     * @param <T> the type of the {@link ThreadLocal}
     * @return an interceptor that changes threadLocal to value during {@link Interceptor#invoke(Method, Object[], Invocation)}
     */
    static <T> Interceptor threadLocal(final ThreadLocal<T> threadLocal, final @Nullable T value) {
        Objects.requireNonNull(threadLocal);
        return (method, arguments, invocation) -> {
            final @Nullable T oldValue = threadLocal.get();
            threadLocal.set(value);
            try {
                return invocation.proceed();
            } finally {
                threadLocal.set(oldValue);
            }
        };
    }

}

File: kotlin/yass/main/ch/softappeal/yass/Interceptor.kt
package ch.softappeal.yass

import java.lang.reflect.*
import java.lang.reflect.Proxy.*

@MustBeDocumented
@Retention(AnnotationRetention.SOURCE)
@Target(AnnotationTarget.FUNCTION)
annotation class OnlyNeededForJava

typealias Invocation = () -> Any?

/** Should return invocation(). */
typealias Interceptor = (method: Method, arguments: List<Any?>, invocation: Invocation) -> Any?

val DirectInterceptor: Interceptor = { _, _, invocation -> invocation() }

fun compositeInterceptor(first: Interceptor, second: Interceptor): Interceptor = when {
    first === DirectInterceptor -> second
    second === DirectInterceptor -> first
    else -> { method, arguments, invocation ->
        first(method, arguments) { second(method, arguments, invocation) }
    }
}

@SafeVarargs
fun compositeInterceptor(vararg interceptors: Interceptor): Interceptor {
    var composite = DirectInterceptor
    for (interceptor in interceptors) composite = compositeInterceptor(composite, interceptor)
    return composite
}

internal fun invoke(method: Method, implementation: Any, arguments: List<Any?>): Any? = try {
    method.invoke(implementation, *arguments.toTypedArray())
} catch (e: InvocationTargetException) {
    throw e.cause!!
}

internal fun invoke(interceptor: Interceptor, method: Method, implementation: Any, arguments: List<Any?>): Any? =
    interceptor(method, arguments) { invoke(method, implementation, arguments) }

internal fun args(arguments: Array<Any?>?): List<Any?> =
    if (arguments == null) listOf() else listOf(*arguments)

internal fun <C : Any> proxy(contract: Class<C>, invocationHandler: InvocationHandler): C {
    var proxy: Any? = null
    val objectMethods: Map<Method, (arguments: Array<Any?>?) -> Any?> = mapOf(
        Object::class.java.getMethod("toString") to { _ -> "<yass proxy for '${contract.canonicalName}'>" },
        Object::class.java.getMethod("hashCode") to { _ -> contract.hashCode() },
        Object::class.java.getMethod("equals", Object::class.java) to { arguments -> proxy === arguments!![0] }
    )
    proxy = newProxyInstance(contract.classLoader, arrayOf(contract), InvocationHandler { p, method, arguments ->
        val objectMethod = objectMethods[method]
        if (objectMethod != null) return@InvocationHandler objectMethod(arguments)
        invocationHandler(p, method, arguments)
    })
    @Suppress("UNCHECKED_CAST") return proxy as C
}

@OnlyNeededForJava
@SafeVarargs
fun <C : Any> proxy(contract: Class<C>, implementation: C, vararg interceptors: Interceptor): C {
    val interceptor = compositeInterceptor(*interceptors)
    if (interceptor === DirectInterceptor) return implementation
    return proxy(
        contract,
        InvocationHandler { _, method, arguments -> invoke(interceptor, method, implementation, args(arguments)) }
    )
}

@SafeVarargs
inline fun <reified C : Any> proxy(implementation: C, vararg interceptors: Interceptor): C =
    proxy(C::class.java, implementation, *interceptors)

fun addSuppressed(e: Exception, block: () -> Unit): Unit = try {
    block()
} catch (e2: Exception) {
    e.addSuppressed(e2)
}

fun <T : Any?, B : Any?> threadLocal(threadLocal: ThreadLocal<T>, value: T, block: () -> B): B {
    val oldValue = threadLocal.get()
    threadLocal.set(value)
    try {
        return block()
    } finally {
        threadLocal.set(oldValue)
    }
}

fun <T : Any?> threadLocalInterceptor(threadLocal: ThreadLocal<T>, value: T): Interceptor = { _, _, invocation ->
    threadLocal(threadLocal, value) { invocation() }
}
